modules/apps/asset/asset-display-page-api/src/main/java/com/liferay/asset/display/page/portlet/BaseAssetDisplayPageFriendlyURLResolver.java

```diff
@@ -203,32 +203,42 @@ public LayoutFriendlyURLComposite getLayoutFriendlyURLComposite(
 			groupId, layoutDisplayPageObjectProvider,
 			layoutDisplayPageProvider);
 
-		HttpServletRequest httpServletRequest =
-			(HttpServletRequest)requestContext.get("request");
+		String originalFriendlyURL = _getOriginalFriendlyURL(friendlyURL);
 
-		HttpSession httpSession = httpServletRequest.getSession();
+		String localizedFriendlyURL = originalFriendlyURL;
 
-		Locale locale = (Locale)httpSession.getAttribute(WebKeys.LOCALE);
+		String urlTitle = layoutDisplayPageObjectProvider.getURLTitle(
+			getLocale(requestContext));
 
-		if (locale != null) {
-			String originalFriendlyURL = _getOriginalFriendlyURL(friendlyURL);
+		if (Validator.isNotNull(urlTitle)) {
+			localizedFriendlyURL = getURLSeparator() + urlTitle;
+		}
 
-			String localizedFriendlyURL = originalFriendlyURL;
+		if (!Objects.equals(originalFriendlyURL, localizedFriendlyURL)) {
+			return new LayoutFriendlyURLComposite(
+				layout, localizedFriendlyURL, true);
+		}
 
-			String urlTitle = layoutDisplayPageObjectProvider.getURLTitle(
-				locale);
+		return new LayoutFriendlyURLComposite(layout, friendlyURL, false);
+	}
 
-			if (Validator.isNotNull(urlTitle)) {
-				localizedFriendlyURL = getURLSeparator() + urlTitle;
-			}
+	protected Locale getLocale(Map<String, Object> requestContext) {
```

Comment:
Avoid creating protected methods unless necessary; make this one private.

Also, can you avoid nesting things like this? Flat code is easier to understand than nested code. Can you do something like:
```
if (locale != null) {
    return locale;
}

....
locale = (Locale)httpSession.getAttribute(WebKeys.LOCALE);

if (locale != null) {
    return locale;
}

...
```

Reply:
Forget about the private thing. I see it is used in subclasses.